File: swift/Tests/DirectoryServiceTests.swift
// SPDX-License-Identifier: MIT
import XCTest
@testable import AetherNetProtocol

/// Tests for ``DirectoryService`` — application-layer name resolution added in
/// v1.2.0 (Issue #60). Mirrors the C# DirectoryServiceTests structure.
final class DirectoryServiceTests: XCTestCase {

    private func sampleDescriptor(rootHash: String = "deadbeef") -> ContentDescriptor {
        ContentDescriptor(
            rootHash: rootHash,
            name: "ignored-publisher-hint",
            totalBytes: 1024,
            chunkSizeBytes: 256,
            chunkCount: 4,
            chunkHashes: ["h0", "h1", "h2", "h3"],
            contentType: "audio/flac"
        )
    }

    // MARK: - publish

    func test_publish_storesLocallyAndBroadcastsNamePublish() async throws {
        let sender = FakeMeshSender(localUhid: "publisher")
        sender.addPeer(PeerInfo(uhid: "peer-1", capabilities: 0))
        sender.addPeer(PeerInfo(uhid: "peer-2", capabilities: 0))
        let dir = DirectoryService(sender: sender)

        try await dir.publish(name: "podcast:abc", descriptor: sampleDescriptor(rootHash: "root-abc"))

        // Local resolve hits the catalogue immediately.
        let hit = await dir.resolve(name: "podcast:abc")
        XCTAssertNotNil(hit)
        XCTAssertEqual(hit?.rootHash, "root-abc")

        // Broadcast went out.
        let broadcasts = sender.broadcasts()
        XCTAssertEqual(broadcasts.count, 1)
        XCTAssertEqual(broadcasts.first?.type, .namePublish)
    }

    func test_resolve_localCatalogueHit_returnsImmediately_noQueryBroadcast() async throws {
        let sender = FakeMeshSender(localUhid: "local")
        sender.addPeer(PeerInfo(uhid: "peer-1", capabilities: 0))
        let dir = DirectoryService(sender: sender)

        try await dir.publish(name: "track:xyz", descriptor: sampleDescriptor(rootHash: "root-xyz"))
        sender.clear()

        let hit = await dir.resolve(name: "track:xyz")

        XCTAssertNotNil(hit)
        XCTAssertEqual(hit?.rootHash, "root-xyz")
        XCTAssertTrue(sender.broadcasts().isEmpty, "no NameQuery should be sent on local catalogue hit")
    }

    // MARK: - Inbound NamePublish

    func test_handle_inboundNamePublish_populatesCatalogueAndFiresEvent() async throws {
        let sender = FakeMeshSender(localUhid: "local")
        let dir = DirectoryService(sender: sender)

        actor Box {
            var captured: DirectoryEntryAnnouncedEvent?
            func set(_ ev: DirectoryEntryAnnouncedEvent) { captured = ev }
        }
        let box = Box()
        await dir.setOnEntryAnnounced { ev in
            Task { await box.set(ev) }
        }

        // Build a NamePublish packet from a peer.
        let descriptor = sampleDescriptor(rootHash: "from-peer")
        let payload = NamePublishPayload(name: "reel:hello", descriptor: descriptor, inResponseToQueryId: nil)
        let body = try JSONEncoder().encode(payload)
        let packet = MeshPacket(
            type: .namePublish,
            sourceUhid: "peer-publisher",
            payload: body
        )
        try await dir.handle(packet)

        // Allow callback Task to flush.
        try? await Task.sleep(nanoseconds: 50_000_000)

        // Local catalogue now has the entry.
        let hit = await dir.resolve(name: "reel:hello")
        XCTAssertNotNil(hit)
        XCTAssertEqual(hit?.rootHash, "from-peer")

        // Event fired.
        let captured = await box.captured
        XCTAssertNotNil(captured)
        XCTAssertEqual(captured?.name, "reel:hello")
        XCTAssertEqual(captured?.sourceUhid, "peer-publisher")
        XCTAssertEqual(captured?.descriptor.rootHash, "from-peer")
    }

    // MARK: - Query / Response roundtrip

    func test_handle_queryWithMatchingName_unicastsNamePublishResponse() async throws {
        let holderSender = FakeMeshSender(localUhid: "holder")
        holderSender.addPeer(PeerInfo(uhid: "asker", capabilities: 0))
        let holder = DirectoryService(sender: holderSender)

        try await holder.publish(name: "album:test", descriptor: sampleDescriptor(rootHash: "album-root"))
        holderSender.clear()

        // Build a NameQuery as if from `asker`.
        let queryId = UUID()
        let queryPayload = NameQueryPayload(name: "album:test", queryId: queryId)
        let queryBody = try JSONEncoder().encode(queryPayload)
        let queryPacket = MeshPacket(
            type: .nameQuery,
            sourceUhid: "asker",
            payload: queryBody
        )

        try await holder.handle(queryPacket)

        // Holder unicasts back a NamePublish with inResponseToQueryId set.
        let unicasts = holderSender.unicasts()
        XCTAssertEqual(unicasts.count, 1)
        guard let response = unicasts.first else { return }
        XCTAssertEqual(response.nextHopUhid, "asker")
        XCTAssertEqual(response.packet.type, .namePublish)

        let decoded = try JSONDecoder().decode(NamePublishPayload.self, from: response.packet.payload)
        XCTAssertEqual(decoded.name, "album:test")
        XCTAssertEqual(decoded.descriptor.rootHash, "album-root")
        XCTAssertEqual(decoded.inResponseToQueryId, queryId)
    }

    func test_handle_queryForUnknownName_doesNothing() async throws {
        let sender = FakeMeshSender(localUhid: "local")
        sender.addPeer(PeerInfo(uhid: "asker", capabilities: 0))
        let dir = DirectoryService(sender: sender)

        let queryPayload = NameQueryPayload(name: "nothing-here", queryId: UUID())
        let queryBody = try JSONEncoder().encode(queryPayload)
        let queryPacket = MeshPacket(
            type: .nameQuery,
            sourceUhid: "asker",
            payload: queryBody
        )

        try await dir.handle(queryPacket)

        XCTAssertTrue(sender.unicasts().isEmpty)
        XCTAssertTrue(sender.broadcasts().isEmpty)
    }

    func test_resolve_missAndTimeout_returnsNil() async {
        let sender = FakeMeshSender(localUhid: "local")
        sender.addPeer(PeerInfo(uhid: "peer-1", capabilities: 0))
        let dir = DirectoryService(sender: sender)

        let hit = await dir.resolve(name: "unknown-name", timeout: 0.05)

        XCTAssertNil(hit)
        // A NameQuery WAS broadcast — we tried.
        let broadcasts = sender.broadcasts()
        XCTAssertEqual(broadcasts.count, 1)
        XCTAssertEqual(broadcasts.first?.type, .nameQuery)
    }

    func test_resolve_queryAndAnswerArrives_returnsDescriptor() async throws {
        let sender = FakeMeshSender(localUhid: "local")
        sender.addPeer(PeerInfo(uhid: "peer-1", capabilities: 0))
        let dir = DirectoryService(sender: sender)

        // Kick off resolve in the background.
        async let resolveResult = dir.resolve(name: "podcast:remote", timeout: 2.0)

        // Wait briefly for the NameQuery to be broadcast.
        try? await Task.sleep(nanoseconds: 100_000_000)

        let broadcasts = sender.broadcasts()
        XCTAssertEqual(broadcasts.count, 1)
        guard let queryBroadcast = broadcasts.first else { return }
        XCTAssertEqual(queryBroadcast.type, .nameQuery)

        let queryDecoded = try JSONDecoder().decode(NameQueryPayload.self, from: queryBroadcast.payload)

        // Simulate a peer responding with a NamePublish carrying inResponseToQueryId.
        let descriptor = sampleDescriptor(rootHash: "remote-root")
        let response = NamePublishPayload(
            name: "podcast:remote",
            descriptor: descriptor,
            inResponseToQueryId: queryDecoded.queryId
        )
        let responseBody = try JSONEncoder().encode(response)
        let responsePacket = MeshPacket(
            type: .namePublish,
            sourceUhid: "peer-1",
            payload: responseBody
        )
        try await dir.handle(responsePacket)

        let result = await resolveResult
        XCTAssertNotNil(result)
        XCTAssertEqual(result?.rootHash, "remote-root")
    }

    // MARK: - Listing

    func test_listNames_returnsCatalogueSnapshot() async throws {
        let sender = FakeMeshSender(localUhid: "local")
        let dir = DirectoryService(sender: sender)

        try await dir.publish(name: "a", descriptor: sampleDescriptor(rootHash: "hash-a"))
        try await dir.publish(name: "b", descriptor: sampleDescriptor(rootHash: "hash-b"))
        try await dir.publish(name: "c", descriptor: sampleDescriptor(rootHash: "hash-c"))

        let names = await dir.listNames()

        XCTAssertEqual(names.count, 3)
        XCTAssertTrue(names.contains("a"))
        XCTAssertTrue(names.contains("b"))
        XCTAssertTrue(names.contains("c"))
    }
}

File: tests/AetherNet.Core.Tests/DirectoryServiceTests.cs
// SPDX-License-Identifier: MIT

using AetherNet.Content;
using AetherNet.Content.Models;
using AetherNet.Core.Tests.Fakes;
using AetherNet.Models;
using AetherNet.Protocol;
using Xunit;

namespace AetherNet.Core.Tests;

/// <summary>
/// Tests for <see cref="DirectoryService"/> — application-layer name resolution
/// added in v1.2.0 (Issue #60).
/// </summary>
public class DirectoryServiceTests
{
    private static ContentDescriptor SampleDescriptor(string rootHash = "deadbeef")
    {
        return new ContentDescriptor
        {
            RootHash = rootHash,
            Name = "ignored-publisher-hint",
            TotalBytes = 1024,
            ChunkSizeBytes = 256,
            ChunkCount = 4,
            ChunkHashes = new[] { "h0", "h1", "h2", "h3" },
            ContentType = "audio/flac",
        };
    }

    // Mirrors DirectoryService.HashName — pins the salted-hash value so the 7 language ports must
    // reproduce it exactly (cross-language parity), and proves the plaintext name never hits the wire.
    private const string NameHashSalt = "aether-dir-name-v1:";
    private static string ExpectedHash(string name)
        => System.Convert.ToHexString(System.Security.Cryptography.SHA256.HashData(
               System.Text.Encoding.UTF8.GetBytes(NameHashSalt + name))).ToLowerInvariant();

    // ─── PublishAsync ────────────────────────────────────────────────────

    [Fact]
    public async Task PublishAsync_StoresLocallyAndBroadcastsNamePublish()
    {
        var sender = new FakeMeshSender("publisher");
        sender.AddPeer(new PeerInfo { Uhid = "peer-1" });
        sender.AddPeer(new PeerInfo { Uhid = "peer-2" });
        var dir = new DirectoryService(sender);

        await dir.PublishAsync("podcast:abc", SampleDescriptor("root-abc"));

        // Local resolve hits the catalogue immediately.
        var hit = await dir.ResolveAsync("podcast:abc");
        Assert.NotNull(hit);
        Assert.Equal("root-abc", hit!.RootHash);

        // Broadcast went out.
        Assert.Single(sender.Broadcasts);
        Assert.Equal(PacketType.NamePublish, sender.Broadcasts.First().Type);
    }

    [Fact]
    public async Task ResolveAsync_LocalCatalogueHit_ReturnsImmediately_NoQueryBroadcast()
    {
        var sender = new FakeMeshSender("local");
        sender.AddPeer(new PeerInfo { Uhid = "peer-1" });
        var dir = new DirectoryService(sender);

        await dir.PublishAsync("track:xyz", SampleDescriptor("root-xyz"));
        sender.Clear();

        var hit = await dir.ResolveAsync("track:xyz");

        Assert.NotNull(hit);
        Assert.Equal("root-xyz", hit!.RootHash);
        Assert.Empty(sender.Broadcasts); // no NameQuery sent — local hit
    }

    // ─── Inbound NamePublish ─────────────────────────────────────────────

    [Fact]
    public async Task HandleAsync_InboundNamePublish_PopulatesCatalogueAndFiresEvent()
    {
        var sender = new FakeMeshSender("local");
        var dir = new DirectoryService(sender);

        DirectoryEntryAnnouncedEventArgs? captured = null;
        dir.EntryAnnounced += (_, e) => captured = e;

        // Build a NamePublish packet from a peer.
        var peerSender = new FakeMeshSender("peer-publisher");
        peerSender.AddPeer(new PeerInfo { Uhid = "local" });
        var peerDir = new DirectoryService(peerSender);
        var descriptor = SampleDescriptor("from-peer");
        await peerDir.PublishAsync("reel:hello", descriptor);

        // Take the broadcast and replay it into the local service.
        var broadcast = peerSender.Broadcasts.First();
        broadcast.SourceUhid = "peer-publisher";
        await dir.HandleAsync(broadcast);

        // Local catalogue now has the entry.
        var hit = await dir.ResolveAsync("reel:hello");
        Assert.NotNull(hit);
        Assert.Equal("from-peer", hit!.RootHash);

        // Event fired.
        Assert.NotNull(captured);
        // PRIVACY: the event carries only the salted hash; the plaintext name is not on the wire.
        Assert.Equal(ExpectedHash("reel:hello"), captured!.NameHash);
        Assert.DoesNotContain("reel:hello", System.Text.Encoding.UTF8.GetString(broadcast.Payload));
        Assert.Equal("peer-publisher", captured.SourceUhid);
        Assert.Equal("from-peer", captured.Descriptor.RootHash);
    }

    // ─── Query / Response roundtrip ──────────────────────────────────────

    [Fact]
    public async Task HandleAsync_QueryWithMatchingName_UnicastsNamePublishResponse()
    {
        var holderSender = new FakeMeshSender("holder");
        holderSender.AddPeer(new PeerInfo { Uhid = "asker" });
        var holder = new DirectoryService(holderSender);

        await holder.PublishAsync("album:test", SampleDescriptor("album-root"));
        holderSender.Clear();

        // Build a NameQuery as if from `asker`.
        var queryId = Guid.NewGuid();
        var queryPayload = System.Text.Json.JsonSerializer.SerializeToUtf8Bytes(
            new NameQueryPayload { NameHash = ExpectedHash("album:test"), QueryId = queryId },
            new System.Text.Json.JsonSerializerOptions { PropertyNamingPolicy = System.Text.Json.JsonNamingPolicy.SnakeCaseLower });
        var queryPacket = new MeshPacket
        {
            Type = PacketType.NameQuery,
            SourceUhid = "asker",
            Payload = queryPayload,
        };

        await holder.HandleAsync(queryPacket);

        // Holder unicasts back a NamePublish with InResponseToQueryId set.
        Assert.Single(holderSender.Unicasts);
        var (responsePacket, nextHop) = holderSender.Unicasts.First();
        Assert.Equal("asker", nextHop);
        Assert.Equal(PacketType.NamePublish, responsePacket.Type);

        var responseBody = System.Text.Json.JsonSerializer.Deserialize<NamePublishPayload>(
            responsePacket.Payload,
            new System.Text.Json.JsonSerializerOptions { PropertyNamingPolicy = System.Text.Json.JsonNamingPolicy.SnakeCaseLower });
        Assert.NotNull(responseBody);
        Assert.Equal(ExpectedHash("album:test"), responseBody!.NameHash);
        Assert.Equal("album-root", responseBody.Descriptor.RootHash);
        Assert.Equal(queryId, responseBody.InResponseToQueryId);
    }

    [Fact]
    public async Task HandleAsync_QueryForUnknownName_DoesNothing()
    {
        var sender = new FakeMeshSender("local");
        sender.AddPeer(new PeerInfo { Uhid = "asker" });
        var dir = new DirectoryService(sender);

        var queryPayload = System.Text.Json.JsonSerializer.SerializeToUtf8Bytes(
            new NameQueryPayload { NameHash = ExpectedHash("nothing-here"), QueryId = Guid.NewGuid() },
            new System.Text.Json.JsonSerializerOptions { PropertyNamingPolicy = System.Text.Json.JsonNamingPolicy.SnakeCaseLower });
        var queryPacket = new MeshPacket
        {
            Type = PacketType.NameQuery,
            SourceUhid = "asker",
            Payload = queryPayload,
        };

        await dir.HandleAsync(queryPacket);

        Assert.Empty(sender.Unicasts);
        Assert.Empty(sender.Broadcasts);
    }

    [Fact]
    public async Task ResolveAsync_MissAndTimeout_ReturnsNull()
    {
        var sender = new FakeMeshSender("local");
        sender.AddPeer(new PeerInfo { Uhid = "peer-1" });
        var dir = new DirectoryService(sender);

        var hit = await dir.ResolveAsync("unknown-name", TimeSpan.FromMilliseconds(50));

        Assert.Null(hit);
        // A NameQuery WAS broadcast — we tried.
        Assert.Single(sender.Broadcasts);
        Assert.Equal(PacketType.NameQuery, sender.Broadcasts.First().Type);
    }

    [Fact]
    public async Task ResolveAsync_QueryAndAnswerArrives_ReturnsDescriptor()
    {
        var sender = new FakeMeshSender("local");
        sender.AddPeer(new PeerInfo { Uhid = "peer-1" });
        var dir = new DirectoryService(sender);

        // Start a resolve in the background.
        var resolveTask = dir.ResolveAsync("podcast:remote", TimeSpan.FromSeconds(2));

        // Wait briefly for the NameQuery to be broadcast.
        await Task.Delay(50);

        Assert.Single(sender.Broadcasts);
        var queryBroadcast = sender.Broadcasts.First();
        Assert.Equal(PacketType.NameQuery, queryBroadcast.Type);

        var query = System.Text.Json.JsonSerializer.Deserialize<NameQueryPayload>(
            queryBroadcast.Payload,
            new System.Text.Json.JsonSerializerOptions { PropertyNamingPolicy = System.Text.Json.JsonNamingPolicy.SnakeCaseLower });

        // Simulate a peer responding with a NamePublish carrying InResponseToQueryId.
        var descriptor = SampleDescriptor("remote-root");
        var responsePayload = System.Text.Json.JsonSerializer.SerializeToUtf8Bytes(
            new NamePublishPayload
            {
                NameHash = ExpectedHash("podcast:remote"),
                Descriptor = descriptor,
                InResponseToQueryId = query!.QueryId,
            },
            new System.Text.Json.JsonSerializerOptions { PropertyNamingPolicy = System.Text.Json.JsonNamingPolicy.SnakeCaseLower });
        var responsePacket = new MeshPacket
        {
            Type = PacketType.NamePublish,
            SourceUhid = "peer-1",
            Payload = responsePayload,
        };
        await dir.HandleAsync(responsePacket);

        var result = await resolveTask;
        Assert.NotNull(result);
        Assert.Equal("remote-root", result!.RootHash);
    }

    // ─── Listing ─────────────────────────────────────────────────────────

    [Fact]
    public async Task ListNamesAsync_ReturnsCatalogueSnapshot()
    {
        var sender = new FakeMeshSender("local");
        var dir = new DirectoryService(sender);

        await dir.PublishAsync("a", SampleDescriptor("hash-a"));
        await dir.PublishAsync("b", SampleDescriptor("hash-b"));
        await dir.PublishAsync("c", SampleDescriptor("hash-c"));

        var names = await dir.ListNamesAsync();

        Assert.Equal(3, names.Count);
        Assert.Contains("a", names);
        Assert.Contains("b", names);
        Assert.Contains("c", names);
    }
}
